server/src/main/java/org/candlepin/auth/KeycloakAuth.java

```diff
@@ -0,0 +1,166 @@
+import javax.servlet.ServletResponse;
+import javax.ws.rs.core.Context;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import org.candlepin.auth.permissions.PermissionFactory;
+import org.candlepin.common.exceptions.CandlepinException;
+import org.candlepin.common.exceptions.ServiceUnavailableException;
+import org.candlepin.common.resteasy.auth.AuthUtil;
+
+import org.keycloak.KeycloakSecurityContext;
+import org.candlepin.service.UserServiceAdapter;
+import org.keycloak.adapters.AdapterDeploymentContext;
+
+import org.keycloak.adapters.ServerRequest;
+import org.keycloak.adapters.RefreshableKeycloakSecurityContext;
+import org.keycloak.adapters.KeycloakDeployment;
+import org.keycloak.adapters.KeycloakDeploymentBuilder;
+
+import org.keycloak.common.VerificationException;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.AccessTokenResponse;
+import org.xnap.commons.i18n.I18n;
+
+import javax.inject.Provider;
+/**
+ * KeycloakAuth
+ */
+
+public class KeycloakAuth extends UserAuth implements AuthProvider {
+
+    @Context private ServletRequest servletRequest;
+    @Context private ServletResponse servletResponse;
+
+    protected AdapterDeploymentContext deploymentContext;
+
+    private AccessTokenResponse response = null;
+
+    private KeycloakDeployment kd;
+
+    @Inject
+    public KeycloakAuth(UserServiceAdapter userServiceAdapter, Provider<I18n> i18nProvider,
+        PermissionFactory permissionFactory) {
+        super(userServiceAdapter, i18nProvider, permissionFactory);
+        createKeycloakDeploymentFrom();
+    }
+
+    private void createKeycloakDeploymentFrom() {
+        try {
+
+            InputStream is = loadKeycloakConfigFile();
+            kd = KeycloakDeploymentBuilder.build(is);
+            deploymentContext = new AdapterDeploymentContext(kd);
+
+        }
+        catch (Exception e) {
+
+        }
+
+    }
+
+    private InputStream loadKeycloakConfigFile() throws FileNotFoundException {
+        String filepath = "/etc/candlepin/keycloak.json";
+        return new FileInputStream(filepath);
+
+    }
+
+
+    @Override
+    public Principal getPrincipal(HttpRequest httpRequest) {
+        try {
+            String auth = AuthUtil.getHeader(httpRequest, "Authorization");
+
+            if (!auth.isEmpty()) {
+
+                String[] authArray = auth.split(" ");
+                if (authArray[0].equals("Basic")) {
+                    return null;
+                }
+                else {
+                    Base64 base64Url = new Base64(true);
+                    String[] splitstring = authArray[1].split("\\.");
+                    String base64EncodedBody = splitstring[1];
+                    String body = new String(base64Url.decode(base64EncodedBody));
+                    JSONObject obj = new JSONObject(body);
+                    String tokenType = obj.getString("typ");
```

Comment:
It is typ only after parsing

Reply:
I don't understand. The key should be safe from any of the parsing being done here, as it would only be exposed after converting the decoded base64 to JSON. At which point, we shouldn't be doing any truncating or splitting on the data.

If the key is indeed "typ", is this something we have control over? Is "typ" an abbreviation for something? Why is the 'e' missing from the key name here?

Reply:
Honestly I dont know why its named typ and not type. 
Have a look at the picture which parses the Jwt token.
![image](https://user-images.githubusercontent.com/11596347/61235848-7a97fe80-a704-11e9-91f3-8468cfbcf5cd.png)

Reply:
Yeesh... looks like they wanted to keep every key at three characters in length.

Alright, thanks for the details here. Disregard my request.